### DISCUSSION TITLE
LineZone detection failed with multiple line

### DISCUSSION THREAD

**Discussion by sceddd** [roboflow/supervision#107] [Q&A] [answered]
i found that in LineZone detection, you try to trigger with all the detections allready track by ByteTrack, but it failed when work with multible LineZone base on the logic that check all the tracker one and count them at the end of for loop.

**Comment by sceddd** [comment]
Hello, its me again.I have check your code again and figure out the line zone you use to count object will, also count the object outside the line, the object that don't even come close to the line. Here is the example base on the one on the tutorial video: 
![image](https://github.com/roboflow/supervision/assets/102157918/6349ac68-494d-40fd-a0b2-f371f81f4ed8)
and also figure out that if you use the line that have the same direction with the one you already initialize, those in_count and out_count will return the same.
![image](https://github.com/roboflow/supervision/assets/102157918/a6a8a411-5d45-4dc9-8021-fa355d67d823)
Hope you understand, im try to explain with my spoiled english.

**Reply by maddust** [reply, marked as answer]
Hey @sceddd  there is an issue open that is related to this bug see issue #87 .  I have solved the issue by modifying the logic a bit to check if the anchor is within the line segment.  I Will update the one I shared in issue #87  since I had to add a "grace period" to the code to avoid double counting when a bounding box jumps.

**Reply by maddust** [reply]
@sceddd 
here you go @SkalskiP  will most likely will come up with a better solution but for now this works for me . 

Adjust the grace period if you want more frames to be ignored until the object can be counted again 
def __init__(self,name, start: Point, end: Point, grace_period: int = 30):

Adjust margin: threshold for the trigger to count the object 
def is_within_line_segment(self, point: Point, margin: float = 5) -> bool:


from typing import Dict

import cv2
import numpy as np

from supervision.detection.core import Detections
from supervision.draw.color import Color
from supervision.geometry.core import Point, Rect, Vector
from typing import Optional


class LineZone:
    def __init__(self,name, start: Point, end: Point, grace_period: int = 30):
        self.name = name
        self.vector = Vector(start=start, end=end)
        self.tracker_state: Dict[str, bool] = {}
        self.already_counted: Dict[str, bool] = {}  # Flag for each tracker_id
        self.frame_last_seen: Dict[str, int] = {}  # Last frame each tracker_id was encountered
        self.current_frame: int = 0  # Current frame counter
        self.in_count: int = 0
        self.out_count: int = 0
        self.grace_period = grace_period  # Number of frames to wait before resetting
    def is_within_line_segment(self, point: Point, margin: float = 5) -> bool:
        """
        Check if a point is within the line segment.

        Attributes:
            point (Point): The point to be checked.
            margin (float): Additional margin added to the line segment boundaries.

        Returns:
            bool: True if the point is within the line segment, False otherwise.
        """
        x_within = min(self.vector.start.x, self.vector.end.x) - margin <= point.x <= max(self.vector.start.x, self.vector.end.x) + margin
        y_within = min(self.vector.start.y, self.vector.end.y) - margin <= point.y <= max(self.vector.start.y, self.vector.end.y) + margin

        return x_within and y_within


    def trigger(self, detections: Detections):
        self.current_frame += 1  # Increment frame counter

        for xyxy, _, confidence, class_id, tracker_id in detections:
            # handle detections with no tracker_id
            if tracker_id is None:
                continue

            # we check if the bottom center anchor of bbox is on the same side of vector
            x1, y1, x2, y2 = xyxy
            anchor = Point(x=(x1+x2)/2, y=y2)  # Bottom center point of bounding box

            # Check if anchor is within the line segment
            if not self.is_within_line_segment(anchor):
                continue

            tracker_state = self.vector.is_in(point=anchor)

            # handle new detection
            if tracker_id not in self.tracker_state:
                self.tracker_state[tracker_id] = tracker_state
                self.already_counted[tracker_id] = False  # When the object appears, it has not been counted yet
                self.frame_last_seen[tracker_id] = self.current_frame  # Update last seen frame
                continue

            # handle detection on the same side of the line
            if self.tracker_state.get(tracker_id) == tracker_state:
                continue

            # check if this crossing has already been counted
            if self.already_counted.get(tracker_id, False):  # If it has been counted, we skip this
                # Check if grace period has passed since last encounter
                if self.current_frame - self.frame_last_seen.get(tracker_id, 0) > self.grace_period:
                    self.already_counted[tracker_id] = False  # Reset after grace period
                else:
                    continue

            self.tracker_state[tracker_id] = tracker_state
            self.already_counted[tracker_id] = True  # After counting, we mark this crossing as already counted
            self.frame_last_seen[tracker_id] = self.current_frame  # Update last seen frame

            if tracker_state:
                self.in_count += 1
            else:
                self.out_count += 1

**Reply by sceddd** [reply]
Thank you @maddust, i'm very appreciate it. 
One more question, in object tracking by ByteTracker, i realize that in a crowd or in a traffic jam, the tracking will not working well on it, like if there is a car hidding behind another one, it lead to a problem that the tracker id will be different if the detections.detect it once again. If i'd like to count the objects in the frame, it'll lose bunch of them. Any idea?

**Reply by maddust** [reply]
> Thank you @maddust, i'm very appreciate it. One more question, in object tracking by ByteTracker, i realize that in a crowd or in a traffic jam, the tracking will not working well on it, like if there is a car hidding behind another one, it lead to a problem that the tracker id will be different if the detections.detect it once again. If i'd like to count the objects in the frame, it'll lose bunch of them. Any idea?

that is very subjective of how good your trained model is, if you are working with a default model like yolov8 or others ,they might not be good enough for your particular situation(you will have to train a custom model) also you can play with the threshold settings of the tracker.  the other thing i can suggest is to avoid setting a line or zones in parts where occlusion happens .   i regularly counts traffic and crows of people . for traffic i set lines or zones where there is free flow of traffic .  here is some example of my implementation of supervision and yolov8 [https://youtu.be/dER7_RMAkuU](https://youtu.be/dER7_RMAkuU) hope it helps

**Reply by SkalskiP** [reply]
Hi, @maddust 👋🏻! Thanks a lot for taking the initiative and helping @sceddd. It makes me super happy that we have people like you - active and keen to help others - in the `supervision` community. 💜

As for your solution, it is a lot simpler than the one I proposed. I'm not sure if it will work in all cases. But @maddust and @sceddd feel free to use it until I'll come up with a more robust fix.
